File: LeetCode/3733-length-of-longest-v-shaped-diagonal-segment/solution.cpp
class Solution {
public:
    int lenOfVDiagonal(vector<vector<int>>& grid) {
        int n = grid.size(), m = grid[0].size();
        auto inb = [&](int r, int c){ return 0 <= r && r < n && 0 <= c && c < m; };

        vector<int> dr = {-1, 1, 1, -1};
        vector<int> dc = { 1, 1,-1, -1};
        vector<int> cw = {1, 2, 3, 0};

        vector run2(4, vector(n, vector<int>(m, 0)));
        vector run0(4, vector(n, vector<int>(m, 0)));

        auto fill_runs_for_dir = [&](int d){
            int rstep = -dr[d], cstep = -dc[d];

            vector<pair<int,int>> starts;
            starts.reserve(n*m);
            for (int i = 0; i < n; ++i)
                for (int j = 0; j < m; ++j)
                    if (!inb(i + dr[d], j + dc[d])) starts.emplace_back(i, j);

            for (auto [si, sj] : starts) {
                int i = si, j = sj;
                while (inb(i, j)) {
                    int fi = i + dr[d], fj = j + dc[d];
                    int next2 = inb(fi, fj) ? run2[d][fi][fj] : 0;
                    int next0 = inb(fi, fj) ? run0[d][fi][fj] : 0;
                    int v = grid[i][j];
                    if (v == 2) {
                        run2[d][i][j] = 1 + next0;
                        run0[d][i][j] = 0;
                    } else if (v == 0) {
                        run0[d][i][j] = 1 + next2;
                        run2[d][i][j] = 0;
                    } else {
                        run2[d][i][j] = run0[d][i][j] = 0;
                    }
                    i += rstep; j += cstep;
                }
            }
        };

        for (int d = 0; d < 4; ++d) fill_runs_for_dir(d);

        bool anyOne = false;
        for (auto& row : grid) for (int v : row) if (v == 1) { anyOne = true; break; }
        if (!anyOne) return 0;

        int ans = 1;

        for (int sr = 0; sr < n; ++sr) {
            for (int sc = 0; sc < m; ++sc) {
                if (grid[sr][sc] != 1) continue;

                for (int d1 = 0; d1 < 4; ++d1) {
                    int i = sr + dr[d1], j = sc + dc[d1];
                    if (!inb(i, j) || grid[i][j] != 2) continue;

                    int len = 1;
                    int expected = 0;

                    int d2 = cw[d1];

                    while (true) {
                        ans = max(ans, 1 + len);

                        int ni = i + dr[d2], nj = j + dc[d2];
                        if (inb(ni, nj)) {
                            if (expected == 0) {
                                ans = max(ans, 1 + len + run0[d2][ni][nj]);
                            } else {
                                ans = max(ans, 1 + len + run2[d2][ni][nj]);
                            }
                        }

                        int ni1 = i + dr[d1], nj1 = j + dc[d1];
                        if (!inb(ni1, nj1)) break;
                        int v = grid[ni1][nj1];
                        if (v != expected) break;
                        i = ni1; j = nj1;
                        ++len;
                        expected = (expected == 0 ? 2 : 0);
                    }
                }
            }
        }

        return ans;
    }
};

File: LeetCode/3733-length-of-longest-v-shaped-diagonal-segment/solution.py
class Solution:
    def lenOfVDiagonal(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        # Diagonals: NE, SE, SW, NW
        dirs: List[Tuple[int, int]] = [(-1, 1), (1, 1), (1, -1), (-1, -1)]
        clockwise = [1, 2, 3, 0]  # NE->SE->SW->NW->NE

        def inb(r, c): return 0 <= r < n and 0 <= c < m

        # Alternating tails (no turn), starting with 2 vs 0 at (i,j)
        run2 = [[[0]*m for _ in range(n)] for _ in range(4)]
        run0 = [[[0]*m for _ in range(n)] for _ in range(4)]

        def fill_runs_for_dir(d: int):
            dr, dc = dirs[d]
            rdr, rdc = -dr, -dc  # traverse so forward neighbor is prefilled
            starts = [(i, j) for i in range(n) for j in range(m)
                    if not inb(i+dr, j+dc)]
            for si, sj in starts:
                i, j = si, sj
                while inb(i, j):
                    fr, fc = i+dr, j+dc
                    nr2 = run2[d][fr][fc] if inb(fr, fc) else 0
                    nr0 = run0[d][fr][fc] if inb(fr, fc) else 0
                    v = grid[i][j]
                    if v == 2:
                        run2[d][i][j] = 1 + nr0
                        run0[d][i][j] = 0
                    elif v == 0:
                        run0[d][i][j] = 1 + nr2
                        run2[d][i][j] = 0
                    else:
                        run2[d][i][j] = run0[d][i][j] = 0
                    i += rdr; j += rdc

        for d in range(4):
            fill_runs_for_dir(d)

        # First-leg prefixes along d, anchored by a preceding 1.
        # Valid start is ONLY 1->2 (never 1->0).
        pre2s = [[[0]*m for _ in range(n)] for _ in range(4)]  # end value 2
        pre0s = [[[0]*m for _ in range(n)] for _ in range(4)]  # end value 0

        def fill_first_leg_for_dir(d: int):
            dr, dc = dirs[d]
            starts = [(i, j) for i in range(n) for j in range(m)
                    if not inb(i-dr, j-dc)]
            for si, sj in starts:
                i, j = si, sj
                while inb(i, j):
                    pr, pc = i-dr, j-dc
                    v = grid[i][j]
                    if inb(pr, pc):
                        pv = grid[pr][pc]
                        if v == 2:
                            if pv == 1:
                                pre2s[d][i][j] = 1         # start 1->2
                            elif pre0s[d][pr][pc] > 0:
                                pre2s[d][i][j] = pre0s[d][pr][pc] + 1
                            else:
                                pre2s[d][i][j] = 0
                            pre0s[d][i][j] = 0
                        elif v == 0:
                            if pre2s[d][pr][pc] > 0:      # extend …2 -> 0
                                pre0s[d][i][j] = pre2s[d][pr][pc] + 1
                            else:
                                pre0s[d][i][j] = 0
                            pre2s[d][i][j] = 0
                        else:
                            pre2s[d][i][j] = pre0s[d][i][j] = 0
                    else:
                        pre2s[d][i][j] = pre0s[d][i][j] = 0
                    i += dr; j += dc

        for d in range(4):
            fill_first_leg_for_dir(d)

        # Combine: no-turn and single clockwise turn.
        if not any(1 in row for row in grid):
            return 0
        ans = 1  # any 1 alone

        for d1 in range(4):
            d2 = clockwise[d1]
            dr2, dc2 = dirs[d2]
            for i in range(n):
                for j in range(m):
                    # First leg ends with 2 -> after turn must start with 0
                    r = pre2s[d1][i][j]
                    if r > 0:
                        ans = max(ans, 1 + r)
                        ni, nj = i + dr2, j + dc2
                        if 0 <= ni < n and 0 <= nj < m:
                            ans = max(ans, 1 + r + run0[d2][ni][nj])
                    # First leg ends with 0 -> after turn must start with 2
                    r = pre0s[d1][i][j]
                    if r > 0:
                        ans = max(ans, 1 + r)
                        ni, nj = i + dr2, j + dc2
                        if 0 <= ni < n and 0 <= nj < m:
                            ans = max(ans, 1 + r + run2[d2][ni][nj])
        return ans

